src/coord/src/timestamp.rs

```diff
@@ -861,97 +904,277 @@ impl Timestamper {
+    type Partition = PartitionId;
+
+    fn get_available_partitions(&self) -> Result<Vec<Self::Partition>, anyhow::Error> {
+        // Poll once to clear any extraneous messages on this queue.
+        self.consumer.poll(Duration::from_secs(0));
+
+        match get_kafka_partitions(&self.consumer, &self.topic, Duration::from_secs(30)) {
+            Ok(partitions) => Ok(partitions
+                .into_iter()
+                .map(|pid| PartitionId::Kafka(pid))
+                .collect()),
+            Err(e) => Err(anyhow!(
+                "Unable to fetch kafka metadata for topic {} (source {}): {}",
+                self.topic,
+                e,
+                self.id
+            )),
+        }
+    }
+}
+
+impl SourceFrontierDiscoverer for RtKafkaDiscoverer {
+    type SourceTimestamp = PartitionOffset;
 
-    let mut current_partition_count = 0;
+    fn get_available_frontier(&self) -> Result<Antichain<Self::SourceTimestamp>, anyhow::Error> {
+        // Poll once to clear any extraneous messages on this queue.
+        self.consumer.poll(Duration::from_secs(0));
 
-    while !c.coordination_state.stop.load(Ordering::SeqCst) {
-        match get_kafka_partitions(&consumer, &c.topic, Duration::from_secs(30)) {
+        match get_kafka_partitions(&self.consumer, &self.topic, Duration::from_secs(30)) {
             Ok(partitions) => {
-                // There cannot be more than i32 partitions
-                let new_partition_count: i32 = partitions.len().try_into().unwrap();
-                match new_partition_count.cmp(&current_partition_count) {
-                    cmp::Ordering::Greater => {
-                        let diff = new_partition_count - current_partition_count;
-                        info!(
-                            "Discovered {} new ({} total) kafka partitions for topic {} (source {})",
-                            diff, new_partition_count, c.topic, c.id,
-                        );
-
-                        for partition in current_partition_count..new_partition_count {
-                            c.coordination_state
-                                .coordinator_channel
-                                .send(coord::Message::AdvanceSourceTimestamp(
-                                    coord::AdvanceSourceTimestamp {
-                                        id: c.id,
-                                        update: TimestampSourceUpdate::RealTime(
-                                            PartitionId::Kafka(partition),
-                                        ),
-                                    },
+                // TODO(benesch): Kafka supports fetching these in bulk, but
+                // rust-rdkafka does not. That would save us a lot of requests on
+                // large topics.
+                let frontier: Result<Vec<_>, _> = partitions
+                    .into_iter()
+                    .map(|pid| {
+                        match self.consumer.fetch_watermarks(
```

Comment:
I like how this proposal simplifies the communication path, but I'm worried about two things - mostly related to this bit here

1. Is this generic enough, or does it only work for sources where we can poll for a high watermark? I think sources where we can poll for watermarks are likely to be the exception.
2. For kafka specifically, is there any path to not needing to issue metadata queries on a schedule? Ideally we would move away from that, but I think this proposal ties us closer to it